File: fundamentals/loops_performance_comparision.py
import time

N = 1000000000  # 1 billion iterations

# --- For loop ---
start = time.perf_counter()
sum_for = 0
for i in range(N):
    sum_for += i
end = time.perf_counter()
print(f"For loop:     {end - start:.3f} seconds (sum={sum_for})")

# --- While loop ---
start = time.perf_counter()
sum_while = 0
j = 0
while j < N:
    sum_while += j
    j += 1
end = time.perf_counter()
print(f"While loop:   {end - start:.3f} seconds (sum={sum_while})")

# --- Do-while loop simulation ---
start = time.perf_counter()
sum_do = 0
k = 0
if N > 0:
    while True:
        sum_do += k
        k += 1
        if not (k < N):
            break
end = time.perf_counter()
print(f"Do-while loop: {end - start:.3f} seconds (sum={sum_do})")

File: fundamentals/loops_performance_comparision.c
#include <stdio.h>
#include <time.h>

#define N 1000000000UL  // 1 billion iterations

int main(void) {
    clock_t start, end;
    double cpu_time_used;
    volatile unsigned long sum;  // volatile prevents compiler optimization

    // --- For loop ---
    sum = 0;
    start = clock();
    for (unsigned long i = 0; i < N; i++) {
        sum += i;
    }
    end = clock();
    cpu_time_used = ((double)(end - start)) / CLOCKS_PER_SEC;
    printf("For loop:     %.3f seconds (sum=%lu)\n", cpu_time_used, sum);

    // --- While loop ---
    sum = 0;
    unsigned long j = 0;
    start = clock();
    while (j < N) {
        sum += j;
        j++;
    }
    end = clock();
    cpu_time_used = ((double)(end - start)) / CLOCKS_PER_SEC;
    printf("While loop:   %.3f seconds (sum=%lu)\n", cpu_time_used, sum);

    // --- Do-while loop ---
    sum = 0;
    unsigned long k = 0;
    start = clock();
    if (N > 0) {
        do {
            sum += k;
            k++;
        } while (k < N);
    }
    end = clock();
    cpu_time_used = ((double)(end - start)) / CLOCKS_PER_SEC;
    printf("Do-while loop: %.3f seconds (sum=%lu)\n", cpu_time_used, sum);

    return 0;
}
